lib/domgen/sync/templates/sync_context_impl.java.erb

```diff
@@ -376,6 +380,66 @@ public abstract class <%= data_module.sync.sync_context_impl_name %>
     query.setParameter( 1, mappingSourceCode );
     return query.executeUpdate();
   }
+<% if entity.sync.supports_bulk_sync? -%>
+
+  @java.lang.Override
+  public int bulkCreate<%= fn %>( @javax.annotation.Nonnull final String mappingSourceCode )
+  {
+    final String insertSql = getSqlToDirectlyInsert<%= fn %>( mappingSourceCode );
+    final javax.persistence.Query insertQuery = _entityManager.createNativeQuery( insertSql );
+    insertQuery.setParameter( 1, mappingSourceCode );
+    final int insertionCount = insertQuery.executeUpdate();
+
+    final String updateSql;
+    if ( insertionCount > 0 )
+    {
+      updateSql = getSqlToMarkInserted<%=fn%>AsSynchronized( mappingSourceCode );
+      final javax.persistence.Query updateQuery =_entityManager.createNativeQuery( updateSql );
+      updateQuery.setParameter(1,mappingSourceCode);
+      final int updateCount = updateQuery.executeUpdate();
+
+      if ( updateCount != insertionCount )
+      {
+        LOG.severe( "Bulk addition of <%= fn %> failed, " + insertionCount + " vs " + updateCount +
+          "\nSql for insert: \n" + insertSql + "\nSql for update: \n" + updateSql );
+        throw new RuntimeException("Mismatch between inserted count and update count for <%= fn %>");
```

Comment:
Of concern: runtime exception

Reply:
Typically I would use a IllegalStateException